cogs/commands/quotes.py

```diff
@@ -1,73 +1,92 @@
 import logging
 import re
 from datetime import datetime
-from typing import List
+from enum import Enum
+from typing import Optional
 
 from discord import AllowedMentions
 from discord.ext import commands
 from discord.ext.commands import Bot, Context, Converter
-from discord.ext.commands.converter import clean_content
 from sqlalchemy.exc import SQLAlchemyError
 from sqlalchemy.sql import func
-from sqlalchemy.sql.operators import is_
 from sqlalchemy_utils import ScalarListException
 
-from models import Quote, QuoteOptouts, db_session, quote
+from models import db_session
+from models.quote import *
 from utils import (
     get_database_user,
     get_name_string,
     is_compsoc_exec_in_guild,
     user_is_irc_bot,
 )
-from utils.mentions import Mention, MentionConverter, MentionType, parse_mention
+from utils.mentions import *
 
 LONG_HELP_TEXT = """
 Pull a random quote. Pull quotes by ID using "#ID", by author using "@username", or by topic by entering plain text
 """
 SHORT_HELP_TEXT = """Record and manage quotes attributed to authors"""
 
 
+class QuoteError(Enum):
+    BAD_FORMAT = 1
+    NOT_PERMITTED = (2,)
+    NOT_FOUND = (3,)
+    OPTED_OUT = (4,)
+    DB_ERROR = (5,)
+    NO_OP = 6
+
+
+class QuoteException(Exception):
+    def __init__(
+        self,
+        err: QuoteError,
+        msg=None,
+    ):
+        self.message = (msg,)
+        self.err = err
+
+
 def is_id(string) -> bool:
-    return re.match("^#\d+$", string)
+    if not isinstance(string, str):
+        return False
+    return re.fullmatch("#\d+", string) is not None
+
+
+"""check if mentioned user has opted out"""
 
 
-def user_opted_out(user: Mention, db_session=db_session):
-    # check if mentioned user has opted out
+def user_opted_out(user: Mention, db_session=db_session) -> bool:
     if user.is_id_type():
-        q = (
-            db_session.query(QuoteOptouts)
-            .filter(QuoteOptouts.user_id == user.id)
-            .count()
-        )
+        f = QuoteOptouts.user_id == user.id
     else:
-        q = (
-            db_session.query(QuoteOptouts)
-            .filter(QuoteOptouts.user_string == user.string)
-            .count()
-        )
+        f = QuoteOptouts.user_string == user.string
+
+    q = db_session.query(QuoteOptouts).filter(f).one_or_none()
```

Comment:
Have you tested this? I've never tried something like this in SQLAlchemy, curious if it works as you'd expect.